### DISCUSSION TITLE
RuntimeError: cycle in page tree

### DISCUSSION THREAD

**Discussion by hellohr11** [pymupdf/PyMuPDF#2493] [Q&A]
_**Please provide all mandatory information!**_

## Describe the bug (mandatory)
When open pdf file in multiple functions, sometimes I got runtimeerror: cycle in page tree. 
It doesn't happen to every pdf. Sometimes, when the error happens, if I change the pdf file path from absolute path to relative path, the problem will be gone.

## To Reproduce (mandatory)

```
import fitz
FILENAME='blueprint (1).pdf'
def do_some_analysis(filename,pageno):
    doc=fitz.open(filename)
    page=doc[pageno-1]
    
    
doc=fitz.open(FILENAME)

for i in range(doc.page_count):
    page=doc[i]
    print(i)
    do_some_analysis(FILENAME,i+1)


```

For problems when building or installing PyMuPDF, give the full output of the build/install command so that, for example, all pip/compiler/linker errors/warnings can be seen.

## Expected behavior (optional)
Describe what you expected to happen (if not obvious).

## Screenshots (optional)
For the same file, running the code snippet
when I use absolute path
<img width="371" alt="image" src="https://github.com/pymupdf/PyMuPDF/assets/45674714/f5f3f2a9-f403-4046-825d-6969ff40740d">
when I use relative path
<img width="111" alt="image" src="https://github.com/pymupdf/PyMuPDF/assets/45674714/d8726f66-4d3d-41a7-aad4-f63be8f91bbf">

## Your configuration (mandatory)
 - Operating system, potentially version and bitness
    ubuntu 22.04
 - Python version, bitness
   python 3.10.6
 - PyMuPDF version, installation method (**wheel** or **generated** from source).
   pip install pymupdf
version 1.21.0

For example, the output of `print(sys.version, "\n", sys.platform, "\n", fitz.__doc__)` would be sufficient (for the first two bullets).
3.10.6 (main, May 29 2023, 11:10:38) [GCC 11.3.0] 
 linux 
 
PyMuPDF 1.21.0: Python bindings for the MuPDF 1.21.0 library.
Version date: 2022-11-08 00:00:01.
Built for Python 3.10 on linux (64-bit).
## Additional context (optional)
[blueprint (1).pdf](https://github.com/pymupdf/PyMuPDF/files/11851411/blueprint.1.pdf)

**Comment by JorjMcKie** [comment]
1. In all generality, this problem is no bug, but a file integrity error.
2. I cannot reproduce the problem with your script and file - neither in Windows nor in Linux.
3. I don't understand, why you open the file multiple times. Why don't do hand the document to `do_some_analysis` instead of the file name? IAW I would suggest doing this:

```python
import fitz

FILENAME = "blueprint.1.pdf"


def do_some_analysis(doc, pageno):
    page = doc[pageno - 1]


doc = fitz.open(FILENAME)

for i in range(doc.page_count):
    page = doc[i]
    print(i)
    do_some_analysis(doc, i + 1)
```

**Comment by hellohr11** [comment]
Thank you @JorjMcKie ,

This script is only an example. We are writing a software analysis pdf files. I passed pdf name and page number to each module. Is there any other way to fix it?

Thanks!

**Comment by JorjMcKie** [comment]
As long as I don't have a reproducing file, I cannot say much I am afraid.
The file you gave me is ok at least, and there is no difference beween using its relative or absolute name.
Very probably a file error: the message says that a cyclic reference between objects in the page tree occurs. 

> This script is only an example. We are writing a software analysis pdf files. I passed pdf name and page number to each module. Is there any other way to fix it?

I understand that that code is more symbolic than anything else, that's fine.

**Comment by JorjMcKie** [comment]
Let's convert this to a Discussion under the hypothesis there is no bug but only file errors.
If you can provide a PDF that reproduces an error, we can revert that decision or open another bug issue as required.
